crates/metadata-store/src/metadata_store.rs

```diff
@@ -371,7 +371,13 @@ impl MetadataStoreClient {
 
     /// Deletes the key-value pair for the given key following the provided precondition. If the
     /// precondition is not met, then the operation returns a [`WriteError::PreconditionViolation`].
-    pub async fn delete(
+    ///
+    /// # Safety
+    /// Parts of the system expect that the version of versioned values increases monotonically
+    /// (e.g. the metadata manager when exchanging metadata information, the raft server when
+    /// rejecting write requests early). Therefore, you should only delete key-value pairs if you
+    /// are certain that these values won't be used in the future ever again!
+    pub async unsafe fn delete(
```

Comment:
I'm not pro using `unsafe` for non-memory-safety related fencing although I know that we have used the same technique in other/older parts of the code base. An alternative is to gate this function to be available in tests only or with a feature flag.